### PR TITLE
feat: add opt-in Realtime reconnect and send queue

### PR DESCRIPTION
**Pull request by jbeckwith-oai**
## Summary

Dropped Realtime connections currently require callers to build their own recovery loop. Add `reconnect: true` to the existing `client.realtime.connect` API, defaulting to `false` for compatibility. Recovery keeps one logical connection and runs the application block once, with finite transient-only retries, jittered backoff, refreshed authentication, and cancellation of outstanding socket work.

Optional byte-bounded retention holds outgoing events while recovering. Applications restore session state and call `flush_pending` explicitly; the SDK never automatically replays an uncertain write. Terminal errors distinguish never-attempted events from uncertain delivery without including payloads in their messages. Each sender receives only its own payload and delivery result, and successful writes are acknowledged atomically before clearing their in-flight identity.

The change is confined to maintained Realtime helpers, RBI/RBS, documentation, and focused tests. Transcription and existing-call sideband behavior are unchanged. Reopening a socket does not restore conversation, audio, or tool state.

## Validation

- Full non-Bedrock test suite against the pinned Steady mock: 1,872 tests, 15,764 assertions, zero failures/errors; one Linux-only procfs test skipped on macOS.
- Realtime, Responses WebSocket, and client authentication on Ruby 3.3/3.4/4.0: 288 tests and 1,813 assertions per version.
- Ten randomized reconnect runs: 350 tests and 2,020 assertions.
- Bedrock: 44 tests, 390 assertions, zero failures/errors; one opt-in live AWS test skipped.
- Full lint, formatting/directive checks, Sorbet and RBS validation, and gem build passed; packaged runtime and signatures verified.
- Custom-code budget passed: 3,451 / 4,000 lines, with no budget or generation-ownership changes.
- Two new consecutive clean adversarial-review rounds, with two independent fresh-context reviewers per round, on the unchanged diff. Transport/security review covers credential refresh, endpoint/TLS policy preservation, cancellation, retention accounting, and uncertain-write handling.

The caller-isolation regressions reproduced the original failure and the close/callback edge cases before the fix, then passed after correction. The repair passed three fresh-context adversarial-review rounds, ending with two consecutive clean rounds on the unchanged diff, plus the full test and lint checks above.

### CODE REVIEW

File: lib/openai/helpers/realtime/recovery.rb [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+
+      private def finish_locked(error = nil)
+        return if @state == :closed
+
+        @error = error
+        @state = :closed
+        @waiters.each do |reply, kind|
+          failure = error ||
+            (kind == :read ? nil : state_error("Realtime connection closed before completing the operation."))
+          reply << [failure ? :error : :ok, failure]
```
[the remark this answers is not in the document]
Comment (jbeckwith-oai) [reply lib/openai/helpers/realtime/recovery.rb]: Fixed in 6125c0e6. Waiting sends now retain their own encoded event and are correlated to the active write by reply identity; terminal errors report only that send’s unsent or uncertain payload. Success is acknowledged atomically before clearing delivery identity. The terminal-entry and restoration-callback paths use the same per-send semantics.

Deterministic regressions reproduce the original concurrent-caller failure (including identical payloads) and the close/callback races, failing before the fix and passing afterward. The full suite (1,872 tests), Ruby 3.3/3.4 compatibility checks, lint/type checks, and two consecutive fresh adversarial-review rounds pass.

**Review by markstuart-oai — APPROVED** [review]
Reviewed all 16 changed files at 6125c0e6ca86e8352e9e4b08f72bfc9b052d11cc, including the maintained Realtime/WebSocket ownership paths, examples, and RBI/RBS contracts. No substantive correctness or structural findings. Recovery remains isolated in Realtime and reuses the canonical handshake/authentication and transport helpers; no changed file crosses 1,000 lines. Explicit restoration/flush and terminal handling of uncertain writes are preserved.

Independently reproduced the four caller-isolation/completion regressions using the pre-fix recovery implementation and verified they pass on this head. Fresh Ruby 4.0.6 validation: 35 reconnect tests / 202 assertions passed; the broader Realtime, Responses WebSocket, and client run completed 329 tests / 1,949 assertions with zero failures and two errors in unchanged proxy tests (also reproduced in isolation). Additional cancellation/queue-accounting probes passed. Sorbet and all 1,290 RBS files validated. I did not rerun the full mock-server suite, cross-version matrix, OkTest, or live API tests.
